[DataGrid] Add sort icon for when column is unsorted

m4theushw opened the pull request:
Closes #1076 
Preview: https://deploy-preview-1415--material-ui-x.netlify.app/components/data-grid/demo

This PR changes the usage of the `hideSortIcons` prop. Using `false` will display an intermediary icon when no sort is applied. To keep the old behavior I've added another `unsorted` (default now) value to only show the sort icon when there's a sort applied.

dtassone commented:
Why do we need the 'unsorted' value exactly?  

So does `hideSortIcons` allow to hide the sort icon for a specific column?

m4theushw commented:
> Why do we need the 'unsorted' value exactly?

`unsorted` is to not show the intermediary icon when the column is unsorted. It reproduces the current behavior. Here the age column uses it and the name column is `false`

<img src="https://user-images.githubusercontent.com/42154031/114734673-d0a33700-9d1a-11eb-8829-f738b7fc4449.png" width="400">

> So does hideSortIcons allow to hide the sort icon for a specific column?

Yep

m4theushw commented on packages/grid/_modules_/grid/components/containers/GridRootStyles.ts:
```diff
@@ -121,8 +117,18 @@ export const useStyles = makeStyles(
-        '& .MuiDataGrid-colCellRight .MuiDataGrid-colCellTitleContainer': {
-          justifyContent: 'flex-end',
+        '& .MuiDataGrid-colCellCenter.MuiDataGrid-withColumnMenu .MuiDataGrid-colCellTitleContainer': {
+          transform: 'translateX(-10px)',
```
Another option is to add `position: absolute` to the menu so it's not considered as part of the div when aligning to the center. This alternative also requires to apply a background or the header name will go over the column menu icon.

![image](https://user-images.githubusercontent.com/42154031/114743605-385d8000-9d23-11eb-8e71-44d8a663c70c.png)

dtassone commented:
`hideSortIcons` I think this `unsorted` setting value is really confusing (to me at least).

To summarise: 
- users can set 3 icons, one for each direction and one when the column is unsorted.
- users might want to hide the sort icons for all columns. For this, he can override the icons and put an empty component, or apply `hideSortIcons` to every sortable columns.
- users might want to hide the sort icons for a specific column. Quite specific edge case IMO, but it could do it using `hideSortIcons`
- users might want to hide the `unsorted` icon for a specific column but show it for another one. `hideSortIcons='unsorted'`. TBH I don't that edge case is necessary. I believe that if users want to hide the unsorted icon, they would do it for all columns, it doesn't make sense to do it for one column and not another. It would be misleading. IMHO, I think we should remove the `unsorted` option for the reason above. Also the prop name is not ideal as it seems like a bool value, but it can take `unsorted`.

m4theushw commented:
> users might want to hide the sort icons for all columns. For this, he can override the icons and put an empty component, or apply hideSortIcons to every sortable columns.

If he changes the icon to null/false/undefined the default icon is used. If changed to something like `div` the button is rendered without the icon, but still takes space.

> users might want to hide the unsorted icon for a all columns, in that case, it can just put an empty component in the unsorted icon.

Same problem as above.

> users might want to hide the unsorted icon for a specific column but show it for another one. hideSortIcons='unsorted'. TBH I don't that edge case is necessary. I believe that if users want to hide the unsorted icon, they would do it for all columns, it doesn't make sense to do it for one column and not another. It would be misleading. IMHO, I think we should remove the unsorted option for the reason above. Also the prop name is not ideal as it seems like a bool value, but it can take unsorted.

The idea of controlling if the unsorted icon should be displayed came from [this](https://github.com/mui-org/material-ui-x/issues/1076#issuecomment-782903407) comment. However, instead of changing to an enum and creating a breaking change, I kept the boolean but added the third option. The [Autocomplete API](https://next.material-ui.com/api/autocomplete/#props) uses this approach in `blurOnSelect` and `forcePopupIcon`.

DanailH commented on docs/src/pages/components/data-grid/columns/StylingHeaderGrid.js:
```diff
@@ -7,11 +7,19 @@ const columns = [
     field: 'last',
     headerClassName: 'super-app-theme--header',
     headerAlign: 'center',
+    hideSortIcons: true,
```
I like the idea but I was thinking of a different API. What if we have a prop like `sortType` and based on its value we choose the icon. So for example possible values can be `dec | asc | unsorted | none`?

dtassone commented:
- Remove `unsorted` value
- Sortable columns should be identifiable with the unsorted icon
- Icon components should allow null value.

oliviertassinari commented:
To add a bit to what we have discussed:

- The root pain that #1076 covers is the accessibility **affordance**.
- An new slot for the components prop can solve this issue, it would be rendered for the "unsorted" state on columns that are sortable.
- Is there really a use case for hiding the sort icons? Why not disable the sort on the column in the first place? How about we remove the `hideSortIcons` property? 
- Accepting `null` for a component slot as a way to disable it could be an exception, we have never done this before, we shouldn't generalize it.

dtassone commented on packages/grid/_modules_/grid/components/columnHeaders/GridColumnHeaderItem.tsx:
```diff
@@ -135,12 +140,11 @@ export const GridColumnHeaderItem = ({
         direction={sortDirection}
         index={sortIndex}
-        hide={column.hideSortIcons}
+        hide={column.hideSortIcons || !column.sortable}
```
it doesn't make sense to have an `hide` prop. Might as well do 
```tsx
{column.hideSortIcons || !column.sortable && (
<GridColumnHeaderSortIcon
        direction={sortDirection}
        index={sortIndex}
        />
)}
```

dtassone commented:
- Issue on this one
![image](https://user-images.githubusercontent.com/936978/115879499-04ccd500-a44a-11eb-91fc-14a531260ae0.png)
- Sort icon should be displayed on the right for boolean
![image](https://user-images.githubusercontent.com/936978/115879793-570df600-a44a-11eb-9638-032f1cb3bd98.png)
- Should we show the unsorted icon by default? If so, this icon does not mean unsorted but sorted desc. I would pick either importExport or unfoldMore from the MUI set. We could also apply an hover style which would increase the opacity to make it less intrusive.
- We used to always show the menu icon on top of all icons, as it allows to do any actions, there is a small regression.
now
![image](https://user-images.githubusercontent.com/936978/115882095-b66d0580-a44c-11eb-9902-e6463e5240bc.png)
![image](https://user-images.githubusercontent.com/936978/115881234-d7812680-a44b-11eb-9cb0-b41ec0d5474b.png)
![image](https://user-images.githubusercontent.com/936978/115882591-2f6c5d00-a44d-11eb-9fdb-7dc9879458ec.png)
before
![image](https://user-images.githubusercontent.com/936978/115882264-e3211d00-a44c-11eb-96ef-b18b02028680.png)
![image](https://user-images.githubusercontent.com/936978/115881470-11522d00-a44c-11eb-8df6-a1b7159738b7.png)

as a side note, this column menu demo is really bad. It just shows how to disable it
https://deploy-preview-1415--material-ui-x.netlify.app/components/data-grid/columns/#column-menu

m4theushw commented:
> Issue on this one

Fixed.

> Sort icon should be displayed on the right for boolean
 
Fixed.

> Should we show the unsorted icon by default? If so, this icon does not mean unsorted but sorted desc. I would pick either importExport or unfoldMore from the MUI set. We could also apply an hover style which would increase the opacity to make it less intrusive.

I changed to importExport but I'm still not happy. I think we should only show the icon on hover and use the same icon of the next sort direction. https://github.com/mui-org/material-ui-x/issues/1076#issuecomment-825183525

> We used to always show the menu icon on top of all icons, as it allows to do any actions, there is a small regression.

I related this problem [here](https://github.com/mui-org/material-ui-x/pull/1415#discussion_r613389932). It's a trade-off of fixing the alignment when using `headerAlign=center`. We can add a background but that's a visual breaking change.

dtassone commented:
> I related this problem here. It's a trade-off of fixing the alignment when using headerAlign=center. We can add a background but that's a visual breaking change.

Yes but the tradeoff seems that it hurts more than the issue. Have you tried with a negative margin?

> I changed to importExport but I'm still not happy. I think we should only show the icon on hover and use the same icon of the next sort direction. #1076 (comment)

Well the issue is to provide the ability to show the unsorted icon when the column is not sorted. If we show the next direction then it would be a different spec. It could be a new option on sortable columns. Something like `showNextDirectionIconOnSortableColumns`, but that would be a different feature.
For the current issue, we have 2 default options.
1. we leave the unsorted icon null.
2. we put an unsorted icon and apply a css to show it on hover. 

As the grid should always come with as many feature as possible, I would lean towards 2.
For css I would go with a low opacity to show it in grey as it's just an info, compared to the active sort state in black.

m4theushw commented:
> Yes but the tradeoff seems that it hurts more than the issue. Have you tried with a negative margin?

Same problem. I applied a negative `margin-right` to the `.MuiDataGrid-colCellTitleContainer` class of the demo [here](https://material-ui.com/components/data-grid/demo/). Since the column menu is transparent, everything under it will be shown. 

![image](https://user-images.githubusercontent.com/42154031/115910996-ebc12580-a443-11eb-9923-208ba0d5d98c.png)

> Well the issue is to provide the ability to show the unsorted icon when the column is not sorted. If we show the next direction then it would be a different spec. It could be a new option on sortable columns. Something like showNextDirectionIconOnSortableColumns, but that would be a different feature. Demo candidate in doc? So we show the customizability of the feature or we wait for a user to create the issue.

Yes, for this we added the new icon slot. The problem related in the issue is solved. My point is which icon we would use as the default one. Since we don't have a proper one my idea was to reproduce the same behavior of https://github.com/mui-org/material-ui-x/issues/1076#issuecomment-825183525 here. I just added now the `GridColumnUnsortedIcon` which is a "dynamic icon", showing the up arrow icon or the down arrow icon, depending on the `sortingOrder`. It can be used as the unsorted icon. I'll still have to fix the CSS to increase its alpha when the column is not sorted.

oliviertassinari commented on docs/pages/api-docs/data-grid.md:
```diff
@@ -114,6 +114,7 @@ Api of the `components` props of type `GridSlotsComponent`
 | <span class="prop-name">ColumnMenuIcon</span> | <span class="prop-type">React.ElementType </span> | <span class="prop-type">TripleDotsVerticalIcon</span> | Icon displayed on the side of the column header title to display the filter input component. |
 | <span class="prop-name">ColumnFilteredIcon</span> | <span class="prop-type">React.ElementType </span> | <span class="prop-type">FilterAltIcon</span> | Icon displayed on the column header menu to show that a filer has been applied to the column. |
 | <span class="prop-name">ColumnSelectorIcon</span> | <span class="prop-type">React.ElementType </span> | <span class="prop-type">ColumnIcon</span> | Icon displayed on the column menu selector tab. |
+| <span class="prop-name">ColumnUnsortedIcon</span> | <span class="prop-type">React.ElementType<br>&#124;&nbsp;null </span> | <span class="prop-type">GridColumnUnsortedIconIcon</span> |  Icon displayed on the side of the column header title when unsorted. |
```
```suggestion
| <span class="prop-name">ColumnUnsortedIcon</span> | <span class="prop-type">React.ElementType<br>&#124;&nbsp;null </span> | <span class="prop-type">GridColumnUnsortedIcon</span> | Icon displayed on the side of the column header title when unsorted. |
```

oliviertassinari commented on docs/pages/api-docs/x-grid.md:
```diff
@@ -120,6 +120,7 @@ Api of the `components` props of type `GridSlotsComponent`
 | <span class="prop-name">ColumnMenuIcon</span> | <span class="prop-type">React.ElementType </span> | <span class="prop-type">TripleDotsVerticalIcon</span> | Icon displayed on the side of the column header title to display the filter input component. |
 | <span class="prop-name">ColumnFilteredIcon</span> | <span class="prop-type">React.ElementType </span> | <span class="prop-type">FilterAltIcon</span> | Icon displayed on the column header menu to show that a filer has been applied to the column. |
 | <span class="prop-name">ColumnSelectorIcon</span> | <span class="prop-type">React.ElementType </span> | <span class="prop-type">ColumnIcon</span> | Icon displayed on the column menu selector tab. |
+| <span class="prop-name">ColumnUnsortedIcon</span> | <span class="prop-type">React.ElementType<br>&#124;&nbsp;null </span> | <span class="prop-type">GridColumnUnsortedIconIcon</span> |  Icon displayed on the side of the column header title when unsorted. |
```
```suggestion
| <span class="prop-name">ColumnUnsortedIcon</span> | <span class="prop-type">React.ElementType<br>&#124;&nbsp;null </span> | <span class="prop-type">GridColumnUnsortedIcon</span> | Icon displayed on the side of the column header title when unsorted. |
```

oliviertassinari commented on packages/grid/_modules_/grid/components/containers/GridRootStyles.ts:
```diff
@@ -103,26 +103,40 @@ export const useStyles = makeStyles(
         },
         '& .MuiDataGrid-colCellSortable': {
           cursor: 'pointer',
+          '&:hover .MuiDataGrid-sortIcon': {
+            opacity: 1,
+          },
```
This looks wrong;

- We don't use opacity 1 in the `<Table>` for the hover unsorted state.
- I was also expecting more **isolation** in the style. Why do we specifically target the styles of `GridColumnUnsortedIcon`?

oliviertassinari reviewed (changes requested):
- How about we document this in https://deploy-preview-1415--material-ui-x.netlify.app/components/data-grid/sorting/
- It seems that there are broken styles. I don't understand why we use a translateX. This is counterintuitive to me.

dtassone commented:
As discussed 
1. let's restore the alignment issue
2. fix the opacity

m4theushw commented on packages/grid/_modules_/grid/components/columnHeaders/GridColumnHeaderItem.tsx:
```diff
@@ -140,15 +144,12 @@ export const GridColumnHeaderItem = ({
-        hide={column.hideSortIcons}
-      />
       <ColumnHeaderFilterIcon counter={filterItemsCounter} />
+      {column.sortable && !column.hideSortIcons && (
```
I had to invert the filter icon and the sort icon or it gonna have an empty space when the column is unsorted, but there's an active filter.

![image](https://user-images.githubusercontent.com/42154031/116129315-5a022400-a6a0-11eb-8975-1673f377e5e3.png)

m4theushw commented on packages/grid/_modules_/grid/components/columnHeaders/GridColumnHeaderSortIcon.tsx:
```diff
@@ -8,51 +8,48 @@ import { GridApiContext } from '../GridApiContext';
+  } else if (direction === 'desc') {
+    Icon = icons!.ColumnSortedDescendingIcon!;
+  }
+  return <Icon fontSize="small" className="MuiDataGrid-sortIcon" />;
```
Added `fontSize` to match with `ColumnHeaderFilterIcon`

oliviertassinari commented on packages/grid/_modules_/grid/components/containers/GridRootStyles.ts:
```diff
@@ -102,27 +102,34 @@ export const useStyles = makeStyles(
           cursor: 'pointer',
         },
-        '& .MuiDataGrid-sortIcon': {
-          fontSize: 18,
```
It looks like we should add this style back. The icon should be 18px, not 20px wide.

oliviertassinari replied:
Well depends on what you read. it feels like 20px in https://material.io/components/data-tables#specs. In https://material-ui.com/components/tables/ we use 18px. On https://search.google.com/, it's 16px:

<img width="244" alt="Screenshot 2021-05-02 at 21 15 21" src="https://user-images.githubusercontent.com/3165635/116824633-8bd92600-ab8b-11eb-9c16-1013a120ac3b.png">

m4theushw replied:
I changed to `font-size: inherit` to not have a specific size for just one icon. They appear to have the same dimensions.

![image](https://user-images.githubusercontent.com/42154031/116894109-119ab700-ac08-11eb-83fa-53d1a906c0ea.png)

oliviertassinari replied:
I see, we rely on IconButton that has font-size 18px on size small. Ok looks good.

oliviertassinari reviewed (approved):
Outside of the font size regression?, it looks great.

dtassone commented:
There are also the issues with the icon on top of the other mentioned 
https://github.com/mui-org/material-ui-x/pull/1415#issuecomment-825681838

oliviertassinari commented:
@dtassone I don't understand. What's the issue? Could you provide enough information on how to reproduce bugs when you report them?

dtassone commented:
> > There are also the issues with the icon on top of the other mentioned #1415 (comment)
> 
> @dtassone I don't follow. What's the issue? Could you provide enough information on how to reproduce bugs when you report them?

As mentioned in the link I put, if you reduce the size of the columns then the icons overlap each others
- We used to always show the menu icon on top of all icons, as it allows to do any actions, there is a small regression.
now
![image](https://user-images.githubusercontent.com/936978/115882095-b66d0580-a44c-11eb-9902-e6463e5240bc.png)
![image](https://user-images.githubusercontent.com/936978/115881234-d7812680-a44b-11eb-9cb0-b41ec0d5474b.png)
![image](https://user-images.githubusercontent.com/936978/115882591-2f6c5d00-a44d-11eb-9fdb-7dc9879458ec.png)
before
![image](https://user-images.githubusercontent.com/936978/115882264-e3211d00-a44c-11eb-96ef-b18b02028680.png)
![image](https://user-images.githubusercontent.com/936978/115881470-11522d00-a44c-11eb-8df6-a1b7159738b7.png)

m4theushw commented:
> As mentioned in the link I put, if you reduce the size of the columns then the icons overlap each others

@dtassone That's a side effect of removing `overflow: hidden` from `.MuiDataGrid-colCellTitleContainer`. It was removed because the value in the badge was being cut when reducing the size of the column. I can't think of a way to stack the column menu without adding the background.

With `overflow: hidden`:

<img width="137" alt="Screenshot 2021-02-21 at 19 20 02" src="https://user-images.githubusercontent.com/3165635/108634421-cfbbfa80-7479-11eb-887f-2d07ecdf3a23.png" style="max-width:100%;">

oliviertassinari commented:
> As mentioned in the link I put

The only link I can find is https://deploy-preview-1415--material-ui-x.netlify.app/components/data-grid/columns/#column-menu. I personally can't see any issue.

dtassone commented:
try overriding the style below.
```tsx
.MuiBadge-anchorOriginTopRightRectangle {
    top: 0;
    right: 0;
    transform: scale(1) translate(30%, -50%);  <= the fix is here, change from 50 to 30%.
    transform-origin: 100% 0%;
}
```

m4theushw commented:
> try overriding the style below.

@dtassone Same problem. 

![image](https://user-images.githubusercontent.com/42154031/116891376-009c7680-ac05-11eb-881c-9439642f5afe.png)

If we invert the horizontal anchor of the `Badge` we can the solve the regression where the sort icon overlaps the column menu.

![image](https://user-images.githubusercontent.com/42154031/116892129-e0b98280-ac05-11eb-9407-b0ab35924559.png)

dtassone commented:
> > try overriding the style below.
> 
> @dtassone Same problem.

you have to restore the overflow and your changes

oliviertassinari commented:
> If we invert the horizontal anchor of the Badge we can the solve the regression where the sort icon overlaps the column menu.

It does feel better too. We would only need to make sure it works if the text alignment is reversed. Otherwise, we could reduce the padding of `.MuiDataGrid-colCell` and add some of it on `.MuiDataGrid-colCellTitleContainer`, so that the overflow hidden doesn't hide the bade.

DanailH commented:
@m2mathew you would need to accept the new screenshots to fix the build 😉

WilliamZimmermann commented:
Hello guys. Finally, the feature to "add sort icon for when column is unsorted" (for <TableSortLabel> component) is not implemented? Because I can't see in the docs (for previous version [4] or the new version [5]) any references to it. If it is not implemented and will not be implemented soon, do you have any suggestion of workaround? Thx!
